plutus-scb/src/Cardano/Protocol/Socket/Server.hs

```diff
@@ -0,0 +1,202 @@
+         some new block is pushed. -}
+      nextState :: Offset
+                -> IO (Either (ChainSync.ServerStNext Block Tip IO ())
+                              (IO (ChainSync.ServerStNext Block Tip IO ())))
+      nextState offset = do
+        st <- readMVar state
+        return $ if offset `lowerThanSlot` st
+                 then Right $ do
+                                st' <- whileT (offset `lowerThanSlot`) state
+                                return $ rollForward st' offset
+                 else Left  $ rollForward st offset
+
+      -- Send the new block to the client.
+      rollForward :: AppState -> Integer -> ChainSync.ServerStNext Block Tip IO ()
+      rollForward st offset =
+        ChainSync.SendMsgRollForward
+          (reverse (st ^. (chainState . chainNewestFirst)) !! (fromIntegral offset))
+          (head    (st ^. (chainState . chainNewestFirst)))
+          (ChainSync.ChainSyncServer (idleState (offset + 1)))
+
+      -- Predicate that selects the slots that have not been seen at the given offset.
+      lowerThanSlot :: Integer -> AppState -> Bool
+      lowerThanSlot offset st =
+        length (st ^. chainState . chainNewestFirst) <= (fromIntegral offset)
+
+      {- Intersect client provided state with the server state and find the
+         best slot to start streaming from. -}
+      intersectState :: ChainState
+                     -> [Point Block]
+                     -> ChainSync.ServerStIntersect Block Tip IO ()
+      intersectState st clientPts =
+        let pt = listToMaybe $ intersect
+                   (getChainPoints st)
+                   clientPts
+        in case pt of
+             Nothing ->
+               ChainSync.SendMsgIntersectNotFound
+                 (head $ view chainNewestFirst st)
+                 -- No intersection found. Resume from origin.
+                 (ChainSync.ChainSyncServer $ idleState 0)
+             Just pt' ->
+               ChainSync.SendMsgIntersectFound
+                 pt'
+                 (head $ view chainNewestFirst st)
+                 (ChainSync.ChainSyncServer $ idleState $ pointOffset pt')
+
+-- | Given a `Point` find its offset into the chain.
+pointOffset :: Point Block
+            -> Integer
+pointOffset pt =
+  case pointSlot pt of
+    Origin        -> 0
+    At (SlotNo s) -> fromIntegral s
+
+-- | Currently selects all points from the blockchain.
+getChainPoints :: ChainState -> [Point Block]
+getChainPoints st =
+  zipWith mkPoint
+    [(st ^. currentSlot) .. 0]
+    (st ^. chainNewestFirst)
+  where
+    mkPoint :: Slot -> Block -> Point Block
+    mkPoint (Slot s) block =
+      Point (At (OP.Block (SlotNo $ fromIntegral s)
+                          (blockId block)))
+
+{- | Wait until a predicate validates the shared data. Used to select
+     the server state if there are new blocks available. -}
+whileT :: (a -> Bool) -> MVar a -> IO a
+whileT p ma = do
+    a <- readMVar ma
+    if p a
+    then do
+      threadDelay 500000
```

Comment:
I know that the unit of `threadDelay`'s argument is microseconds because I recently looked at it, but for readability I think it's nice to use [`Data.Time.Units`](https://hackage.haskell.org/package/time-units-1.0.0/docs/Data-Time-Units.html)

```haskell
let delay = 5 :: Second
threadDelay (toMicroseconds delay)
```

Reply:
It is actually 500ms, but it only brings more evidence to what you are saying. I have added a 500ms delay, following your recipe.